src/Pact/Runtime/Typecheck.hs

```diff
@@ -0,0 +1,177 @@
+-- Output checking -- app return values -- left to static TC.
+-- Native funs not checked here, as they use pattern-matching etc.
+typecheck :: [(Arg (Term Name),Term Name)] -> Eval e ()
+typecheck ps = foldM_ tvarCheck M.empty ps where
+  tvarCheck m (Arg {..},t) = do
+    r <- typecheckTerm _aInfo _aType t
+    case r of
+      Nothing -> return m
+      Just (v,ty) -> case M.lookup v m of
+        Nothing -> return $ M.insert v ty m
+        Just prevTy | prevTy == ty -> return m
+                    | otherwise ->
+                        evalError (_tInfo t) $ "Type error: values for variable " <> pretty _aType <>
+                        " do not match: " <> pretty (prevTy,ty)
+
+-- | Typecheck a def against a declared fun type.
+typecheckDef :: Def Ref -> FunType (Term Name) -> FunType (Term Name) -> Eval e ()
+typecheckDef def defTy funTy = validateArgCount >> tcReturn >> tcArgs
+  where
+    defArgCount = length $ _ftArgs defTy
+    ftyArgCount = length $ _ftArgs funTy
+    validateArgCount
+      | defArgCount == ftyArgCount = return ()
+      | otherwise = evalError' def $ "Bad arg count in '" <> pretty (_dDefName def) <>
+                    "', expected " <> pretty ftyArgCount <>
+                    ", found " <> pretty defArgCount
+    tcReturn = tc def (_ftReturn defTy) (_ftReturn funTy)
+    tcArgs = zipWithM_ tcArg (_ftArgs defTy) (_ftArgs defTy)
+    tcArg a specA = tc a (_aType a) (_aType specA)
+    tc i ty specTy = unless (ty `canUnifyWith` specTy) $
+      typecheckFailed i ty specTy
+
+
+-- | Typecheck applied args against a FunType from some Def.
+typecheckArgs
+  :: HasInfo i
+  => i
+  -> DefName
+  -> FunType (Term Name)
+  -> [Term Name]
+  -> Eval e ()
+typecheckArgs i defName ft' as' = do
+  let params = _ftArgs ft'
+  when (length params /= length as') $
+    evalError' i $ pretty defName <> ": Incorrect number of arguments (" <>
+      pretty (length as') <> ") supplied; expected " <> pretty (length params)
+  typecheck (zip params as')
+
+
+typecheckFailed :: (HasInfo i, Pretty a, Pretty b) => i -> a -> b -> Eval e c
+typecheckFailed i found spec = evalError' i $
+  "Type error: expected " <> pretty spec <> ", found " <> pretty found
+
+-- | 'typecheckTerm i spec t' checks a Term 't' against a specified type 'spec'.
+-- Returns `Nothing` on successful check against concrete/untyped,
+-- or `Just` a pair for successful check against a type variable, where
+-- the pair is the type variable itself and the term type.
+typecheckTerm :: forall e . Info -> Type (Term Name) -> Term Name
+       -> Eval e (Maybe (TypeVar (Term Name),Type (Term Name)))
+typecheckTerm i spec t = do
+
+  ty <- case typeof t of
+    Left s -> evalError i $ "Invalid type in value location: " <> pretty s
+    Right r -> return r
+
+  let
+
+    tcFail found = typecheckFailed i found spec
+
+    tcOK = return Nothing
+
+    -- | check container parameterized type.
```

Comment:
What does "container parametrized type" mean?

Reply:
List or object. I'll update for clarity